implementation/src/main/java/io/smallrye/config/ConfigValueMapView.java

```diff
@@ -0,0 +1,147 @@
+package io.smallrye.config;
+
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The ConfigValueMapView is view over a Map of String configs names and ConfigValue value.
+ * <p>
+ *
+ * Use this to wrap the ConfigValue map and expose it where a Map of String name and String value is required.
+ */
+public class ConfigValueMapView implements Map<String, String> {
+    private final Map<String, ConfigValue> delegate;
+
+    ConfigValueMapView(final Map<String, ConfigValue> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public int size() {
+        return delegate.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(final Object key) {
+        return delegate.containsKey(key);
+    }
+
+    @Override
+    public boolean containsValue(final Object value) {
+        return values().contains(value);
+    }
+
+    @Override
+    public String get(final Object key) {
+        final ConfigValue configValue = delegate.get(key);
+        return configValue != null ? configValue.getValue() : null;
+    }
+
+    @Override
+    public String put(final String key, final String value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public String remove(final Object key) {
```

Comment:
If you extend `AbstractMap`, you don't have to provide method bodies for these mutation operations.

Reply:
But AbstractMap has actual implementations for these. Don't we want to make this read only and mark this methods with UnsupportedOperationException?

Reply:
We can; however, the abstract implementations ultimately result in throwing that exception if the underlying collection is not writable.  Code which tries to write to these collections is already erroneous so it doesn't really matter if it misbehaves efficiently or inefficiently.

Reply:
Ok. Let me change that.

Reply:
Actually, thinking a little better about this, I think it becomes somehow confusing. Extending AbstractMap, will perform the operations on the instance class and not the delegate Map view we are exposing. So, if you do something like:

```java
map.size() == 3
map.remove(x) // x exists, but the method returns null
map.size() == 3 and not 2
```
If I understood correctly, your suggestion was to remove these methods and not provide the implementation bodies. To support this properly we would still need to provide the right delegates here. Is that ok?

Reply:
If the contract of `Map` and its entry set are correctly implemented, if `x` exists it should throw `UnsupportedOperationException` and if it doesn't exist, `null` is returned.  This is correct behavior AFAICT.  Whether or not `AbstractMap` is used (or `AbstractCollection`/`AbstractSet` in the case of the entry sets), any implementation must obey the contract or incorrect/unexpected behavior will result.

Reply:
Sorry, maybe I didn't explain myself correctly. That is true.

In this case, for our implementation the `.remove` is called on the instance class of `ConfigValueMapView extends AbstractMap` if we don't override it, while for instance `.size` is called on our delegate.

So if we omit the `remove`, `put` ,`putAll` and `clear` operations, they will be invoked on the instance Map (without delegation) they will not give you a `UnsupportedOperationException` but won't perform the expected action either, because they are not acting on the delegate map.

Reply:
> So if we omit the `remove`, `put` ,`putAll` and `clear` operations, they will be invoked on the instance Map (without delegation) they will not give you a `UnsupportedOperationException` but won't perform the expected action either, because they are not acting on the delegate map.

This is only partially correct.  Yes, they will be invoked on the instance map without delegation.  But they *will* give you `UnsupportedOperationException`.  Not because the delegate gives it, but because the abstract implementations will result in it.

Reply:
Apparently not. Let me push so we can see it in the CI build.

Reply:
Or maybe I'm forgetting something.

Reply:
The only reason it wouldn't work should be if something is not implemented correctly (the entry set perhaps).

Reply:
It is mostly on the input parameters. I'm still thinking about `Collections.unmodifiableX`, which gives you a `UnsupportedOperationException` straight away.

Like this, if you pass an empty map to `putAll`, it will work because nothing is being added and only if there is something to add,`put` is called and then the `UnsupportedOperationException` is thrown.

Same for `remove` if you try to remove something that is not there you get the null and not the `UnsupportedOperationException`.

I would prefer for this to behave as a `Collections.unmodifiableMap`, because it may give you the false sense of these operations to be possible, depending on what you use to call them.

Reply:
If one chooses to ignore method contracts, whatever happens is probably their own fault. :-)  But if you want exactly `unmodifiableMap` behavior, why not wrap the instance with `unmodifiableMap`?  It's still substantially less code that way.

Reply:
Yes, I thought about that. It was to avoid yet another wrapper, but if we are fine with that, I'll do it :)

Reply:
Ok, check if everything is ok now :)